# PHARE

```
            *           /
  \         |         /
    \      / \      /
      \  /_____\  /
      /  |. | ||  \
    /  |_|:_|_||_|  \
  /     \.      /     \
         |;:   |        \
         |;:.  |          
         |;:  .|
         |;: . | 
         |;:   |            
         |;: , |-~^~ ~^             . -~^~ ~^~ -^~                        
^ -~^~  _|;:  _|\_      -~^~ ~^~ -^~    
 ^    /^    .| |. |-_              ^          ~
     |    ^     ^    |   ~              -
  ~  |       *   |   \-          ~             .
     ===-'-.~~'-~.~-'""~ 



```

# Finding _**P**. falciparum_ **ha**plotypes with **re**sistance mutations in polyclonal infections

A fast and simple pipeline to detect haplotypes in Oxford Nanopore MinION data from multiclonal samples.

![an example of the PHARE pipeline output](resources/pipeline_example.png)

## Requirements:
- A conda/mamba installation is required to run the pipeline.

## Applications:
The PHARE pipeline was developed and tested using Oxford Nanopore Technologies MinION sequences of _Plasmodium falciparum_. The reads can be basecalled using the "Super High Accuracy" basecaller model in guppy or Dorado. The use of the "fast" basecalling model is strongly discouraged, since it will hamper the pipelines ability to accurately detect SNP sites.

The pipeline will likely work with other long-read sequencers, however this was not validated. Short-read sequencing datasets such as generated by Illumina can be analyzed, however only haplotypes of short genes or gene segments can be generated, limiting the informative value.


## Getting Started:

### 1. Download the following files from NCBI GeneBank:
- jsonl file from genbank gene browser
- fasta file with the whole gene

To download these files go to a GenBank gene Database entry such as https://www.ncbi.nlm.nih.gov/gene/2655294 and click on "Download Datasets". Make sure that "Gene Sequence" is checked and click download. In the downloaded folder, locate gene.fna and data_report.jsonl, give them appropriate names and copy them into the resources directory of the pipeline. Then add these files to the corresponding gene in the config file (step 2).

For the _Plasmodium falciparum_ genes _dhps_, _dhfr_ and _K13_ these files can be found in the resources folder.

### 2. Adjust the Configuration

- Add a sample_list.xlsx file to the config folder. Two columns are required: 
    * "barcode": a number representing the barcode used for ONT sequening
    * "sample": a sample name to be displayed in the plot

- Modify the configuration file under configs/config.yaml according to your needs. Detailed descriptions of the parameters are provided in the config files itself or in the publication describing this pipeline. The following parameters can be adjusted:
    + sample_dir: path to the folder containing the basecalled nanopore reads in .fastq format.
    + minqual: the minimum quality of reads and nucleotides. This depends on your dataset.
        * For a dataset with a low coverage, a lower quality score should be used. It also depends on the number of reads retained after running the pipeline. 
        * The lower minimum quality should then be compensated by an increased threshold in the subsequent analysis in R
    + SNP_selection_cut-off: Minimum % of sequencing reads with SNP at a site.
    + coverage_minimum: Minimum % of reads covering a nucleotide position considered a SNP.
    + minimum_haplotype_frequency: Minimum % of reads with certain haplotype to filter out low abundance haplotypes that are considered noise.
    + minimal_number_of_reads_per_sample: Minimum number of reads for a sample to be included in the finaly plot.
    
    + For each gene which should be analyzed, add the following under the "genes" parameter
        * `<gene name>`:
            - targetlen: the length of the amplicon which was sequenced
            - minlen: the minimum length of reads to be considered for the pipeline (default: targetlen + 250)
            - maxlen: the maximum length of reads to be considered for the pipeline (default: targetlen - 50)
            - generef: path to the previously downloaded .fasta file in resources
            - genebank_ref: path to the previously downloaded .jsonl file in resources
            - contig: reference name/contig. This needs to be ***exactely the same as the heading of the fasta file***
        


### 3. Execute the Snakemake pipeline
- The pipline requires one .fastq file per sample named i.e. "barcode13". The barcode number must be included at the end of the sample name and correspond to the barcode defined in sample_list.xlsx.
- Navigate to the root directory of the pipeline (`cd <your_path>/PHARE`) and run snakemake like this: `snakemake --cores all --use-conda`






<!-- 
See the comments in the R file for details about all paramterers
Most importantly though, a threshold has to be set to differentiate between noise and real haplotypes. It is the minimum fraction of reads as part of all reads in a sample that have to be of a certain haplotype, for this haplotype to be considered not noise. (noise = sequencing errors)

The R script consists of the following functions:

- a function to do the initial setup. all variables are stored in a list for easy access (and to simulate a kind of class like structure)
- a function to read all files in a path
    * the barcode number is inferred from the file name! it assumes one to three numbers in the file name: first, the barcode/source, then the subset and lastly the replicate number for in silico replicates
- a function to join the sample names to the data frame and calculate the total number of reads
- a function to create snp/roi dataframe and a reference sequence
    + gb_report: a reportfile downloaded from genbank in .json format, which corresponds to the reference sequence. it is used to calculate exon and intron positions.
    + hapl: the dataframe with all the snps => used to create the output df
- filter duplicate snps
    + if we have a duplicate snp (same aa position but different nucleotide)
- translate nucleotide snps to amino acids
- calculate the frequency of the aa variants
- apply a treshold and then recalculate haplotypes and their relative frequencies for all the haplotypes which still exist (are not considered noise)
- make the dataframe necessary for the plot
- create a plot
- store a plot -->